Public/Src/Sandbox/MacOs/Interop/Sandbox/Data/Trie.hpp

```diff
@@ -0,0 +1,366 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#ifndef Trie_hpp
+#define Trie_hpp
+
+#include "SandboxedProcess.hpp"
+
+/*!
+ * A node in a Trie.
+ * Only accessible to its friend class Trie.
+ */
+class Node
+{
+private:
+
+    friend class Trie;
+
+    _Atomic static uint s_numUintNodes;
+    _Atomic static uint s_numPathNodes;
+
+    /*!
+     * The value 65 is chosen so that all ASCII characters between 32 (' ') and 122 ('z')
+     * get a unique entry in the 'children_' array.  The formula for mapping a character
+     * ch to an array index is:
+     *
+     *   toupper(ch) - 32
+     */
+    static const uint s_pathNodeChildrenCount = 65;
+
+    /*! For 10 digits */
+    static const uint s_uintNodeChildrenCount = 10;
+
+    /*! Arbitrary value */
+    SandboxedProcess *record_;
+
+    /*! The length of the 'children_' array (i.e., the the number of allocated nodes) */
+    uint childrenLength_;
+
+    /*! Pre-allocated pointers to all possible children nodes. */
+    Node **children_;
+
+    uint length()     const { return childrenLength_; }
+    Node** children() const { return children_; }
+    
+    static Node* createUintNode() { return new Node(s_uintNodeChildrenCount); }
+    static Node* createPathNode() { return new Node(s_pathNodeChildrenCount); }
+    
+public:
+    
+    Node() = delete;
+    Node(uint numChildren);
+    ~Node();
+};
+
+// ================================== class Trie ==================================
+
+/*!
+ * A thread-safe, lock-free, dictionary implementation.
+ *
+ * Only 2 types of keys are allowed: (1) an unsigned integer, and (2) an ascii path.
+ *
+ * A value must be a pointer to an arbitrary OSObject.  Once an OSObject is added
```

Comment:
These comments about `OSObject` and `OSDictionary` are probably not relevant any more (just noticing, nothing needs to be done about it now)